### ISSUE TITLE
Feature request: support unix domain socket as listener for MQTT(plain text)

### ISSUE THREAD

**Issue by kushwiz** [vernemq/vernemq#1736]
### To help us save time and help you faster

Before opening a new issue, please make sure that isn't already an open (or
closed/resolved) issue concerning the problem. If unsure if older issues are
about the exact same thing, just open a new issue rather than posting in the old
one - but please reference the possibly related issues in the new issue.

Please do not insert images of text, but insert the text instead.

### Environment

- VerneMQ Version:
- OS:
- Erlang/OTP version (if building from source):
- VerneMQ configuration (vernemq.conf) or the changes from the default
- Cluster size/standalone:

### Expected behavior

- Please describe the expected behaviour.

### Actual behaviour

- Please include minimal steps to reproduce the issue, if possible.

**Comment by ioolkos** [comment]
@kushwiz thanks that's an interesting idea for which I'm at least interested in seeing what the implementation effort would be.

There's a little bit of a discussion here, on adding Unix socket support to Mosquitto: https://github.com/eclipse/mosquitto/pull/467
I'm unclear whether this lead to an actual implementation but it looks like it did not.

In general, I'd welcome if you add a little bit of information why you need this (use case etc).

**Comment by kushwiz** [comment]
My architecture looks this way:

[|PassThru-LoadBalancer|] ---> [|HAProxy| ----> |VerneMQ|]^n

I am using HAProxy just for SSL termination and when using 127.0.0.1 as my backend, I am at the risk of port exhaustion for connections between HAProxy and VerneMQ. Unix sockets should help me bypass this limitation and this would scale better.

**Comment by ioolkos** [comment]
@kushwiz I think this is fairly easy to support and a valid use case.
Note that you can avoid port exhaustion on the HAProxy side with with virtual source IPs.

So from my side: in general, :+1: for implementing UDS support, but low prio.

**Comment by kushwiz** [comment]
@ioolkos thanks for the tip about virtual source IPs. !

**Comment by kushwiz** [comment]
@ioolkos will this be added to low priority list on the features board?

**Comment by ralight** [comment]
Just passing by... Mosquitto does support this now, and you can use its command line clients to test `mosquitto_sub --unix /tmp/your.sock ...`

**Comment by ioolkos** [comment]
@ralight great to know, thanks!

**Comment by ioolkos** [comment]
>  will this be added to low priority list on the features board?

@kushwiz 
yes :)

**Comment by ioolkos** [comment]
@kushwiz though it was fun looking into it & have it working :) but still need to add a couple of stuff, namely session handling & authorization
![unix_sockets](https://user-images.githubusercontent.com/282359/109395607-8a8e4180-792d-11eb-88d6-3224e60bf14c.png)

**Comment by kushwiz** [comment]
@ioolkos Awesome.  Hope this gets released soon 😅

**Comment by kushwiz** [comment]
@ioolkos we have been waiting on this for almost a year now :)

**Comment by kushwiz** [comment]
When we use HAProxy with VerneMQ - the traffic between them remains unencrypted. This will eliminate the requirement of encryption of traffic between the two as unix sockets will remain local to the filesystem.

**Comment by ioolkos** [comment]
@kushwiz a first version for unix domain listeners is now in master. 
Can be configured with `listener.tcp.unix_socket = local:/tmp/my.sock:0` in the `vernemq.conf` file. Improvements will follow (like dynamicall start/stop listener)

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.
